modules/linear_boltzmann_solvers/executors/pi_keigen_smm.cc

```diff
@@ -0,0 +1,1122 @@
+  log.Log() << "Done assembling diffusion system.";
+}
+
+void
+PowerIterationKEigenSMM::Execute()
+{
+  // Start transport power iterations
+  double k_eff_ell = k_eff_;
+  double k_eff_change = 1.0;
+
+  auto phi_ell = phi_old_local_;
+
+  int nit = 0;
+  while (nit < max_iters_)
+  {
+    // Compute the transport 1/k fission source
+    SetLBSFissionSource(phi_ell, /*additive=*/false);
+    Scale(q_moments_local_, 1.0 / k_eff_);
+
+    // Solve some transport inners
+    primary_ags_solver_->Setup();
+    primary_ags_solver_->Solve();
+
+    std::vector<double> phi0;
+    TransferTransportToDiffusion(phi_new_local_, phi0);
+    auto phi0_old = phi0;
+    auto phi0_m = phi0;
+
+    // Update second-moment method data
+    ComputeClosures(psi_new_local_);
+    const auto correction = ComputeSourceCorrection();
+
+    // Start diffusion power iterations
+    double lambda = k_eff_;
+    double lambda_old = k_eff_;
+    double F0_old = lbs_solver_.ComputeFissionProduction(phi_new_local_);
+
+    for (int m = 0; m < accel_pi_max_its_; ++m)
+    {
+      // Define the nodal diffusion 1/lambda fission source.
+      // This returns a local PWLD vector which must still be integrated
+      // against the diffusion test functions.
+      auto Sf = SetNodalDiffusionFissionSource(phi0_m);
+      Scale(Sf, 1.0 / lambda);
+
+      // Solve the diffusion inners. This seems quite wasteful, so
+      // only a single iteration is taken.
+      for (int i = 0; i < 1; ++i)
+      {
+        // Define the nodal diffusion scattering source.
+        // This returns a local PWLD vector which must still be integrated
+        // against the diffusion test functions.
+        auto Ss = SetNodalDiffusionScatterSource(phi0_old);
+
+        // Assemble the normal terms of the RHS of the diffusion solver,
+        // add in the second moment correction, and solve the system
+        auto b = AssembleDiffusionRHS(Sf + Ss);
+        diffusion_solver_->AddToRHS(b + correction);
+        diffusion_solver_->Solve(phi0, /*use_initial_guess=*/true);
+
+        // Bump the new solution to old
+        phi0_old = phi0;
+      }
+
+      // Compute a new diffusion eigenvalue. This requires mapping the
+      // diffusion solution back to the transport discretization.
+      std::vector<double> phi;
+      TransferDiffusionToTransport(phi0, phi);
+      double F0 = lbs_solver_.ComputeFissionProduction(phi);
+      lambda = F0 / F0_old * lambda_old;
+
+      // Check for convergence
+      double lambda_change = std::fabs(lambda / lambda_old - 1.0);
+      if (accel_pi_verbose_)
+        log.Log() << "SMM PI iteration " << m << "  lambda " << lambda << "  change "
+                  << lambda_change << (lambda_change < accel_pi_k_tol_ ? "  CONVERGED" : "");
+
+      // Bump variables for next iteration
+      lambda_old = lambda;
+      F0_old = F0;
+      phi0_m = phi0;
+
+      if (lambda_change < accel_pi_k_tol_)
+        break;
+
+    } // acceleration
```

Comment:
If this is to tell the reader about the acceleration loop, then this should be before the `for` statement.
If this is just copy-paste, maybe even delete this. 

Also, these comments _were_ helpful when for loops were long and people used editors that were not able to go to the matching opening curly brackets. These days we don't really need these, IMO. Maybe, it is time to have discussion about these comments. I can see them a lot in the code. What does @ragusa and @wdhawkins think about those? Useful? Not useful?

Reply:
I am in favor of keeping them. They don't hurt anything IMO and help when the modern IDEs are not available.

Reply:
I agree with @andrsd. Unless it's a very long loop that spans several hundred lines, I don't find the end-of-loop comments very useful and usually remove them during a refactor.

Reply:
This is a topic that we should likely discuss on Wednesday. I prefer to have the end of loop comments, *especially* for deeply nested loops. I am not a fan of things like:
```
            }
          }
        }
      }
    }
  }
}
```
Having comments helps me keep my place when developing, especially when a counter needs to be incremented or something like that. Otherwise, I find myself scrolling up and down to make sure I am in the correct level of nesting for the line of code.

Reply:
Valid point... in cases like that, I can see where end-of-loop comment could be very useful.